drivers/pwm/pwm_gecko.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2021 Sun Amar.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT silabs_gecko_pwm
+
+#include <errno.h>
+#include <drivers/pwm.h>
+#include <dt-bindings/pwm/pwm.h>
+#include "em_cmu.h"
+#include "em_timer.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(pwm_gecko, CONFIG_PWM_LOG_LEVEL);
+
+/** PWM configuration. */
+struct pwm_gecko_config {
+	TIMER_TypeDef *timer;
+	CMU_Clock_TypeDef clock;
+	TIMER_Prescale_TypeDef prescale_enum;
+	uint8_t channel;
+	uint8_t location;
+	uint8_t port;
+	uint8_t pin;
+};
+
+struct pwm_gecko_data {
+	int prescaler;
+};
+
+#define DEV_CFG(dev) \
+	((const struct pwm_gecko_config * const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct pwm_gecko_data *const)(dev)->data)
+
+static int pwm_gecko_pin_set(const struct device *dev, uint32_t pwm,
+			     uint32_t period_cycles, uint32_t pulse_cycles,
+			     pwm_flags_t flags)
+{
+	const struct pwm_gecko_config *cfg = DEV_CFG(dev);
+
+	if (BUS_RegMaskedRead(&cfg->timer->CC[pwm].CTRL,
+		_TIMER_CC_CTRL_MODE_MASK) != timerCCModePWM) {
+
+#ifdef _TIMER_ROUTE_MASK
+		BUS_RegMaskedWrite(&cfg->timer->ROUTE,
+			_TIMER_ROUTE_LOCATION_MASK,
+			cfg->location << _TIMER_ROUTE_LOCATION_SHIFT);
+		BUS_RegMaskedSet(&pwm->timer->ROUTE, 1 << pwm);
+#elif defined(_TIMER_ROUTELOC0_MASK)
+		BUS_RegMaskedWrite(&cfg->timer->ROUTELOC0,
+			_TIMER_ROUTELOC0_CC0LOC_MASK << (pwm * _TIMER_ROUTELOC0_CC1LOC_SHIFT),
+			cfg->location << (pwm * _TIMER_ROUTELOC0_CC1LOC_SHIFT));
+		BUS_RegMaskedSet(&cfg->timer->ROUTEPEN, 1 << pwm);
+#else
+		LOG_DBG("This specific gecko mcu not supported yet");
```

Comment:
Instead of logging, i would prefer a `#error Unsupported device` to get the error during compilation.